Ocean Cubes (#19)
deepesdl/cube-gen: Ocean/README.md

# Polar Datacube Generation

> Information dedicated to the generation of the Ocean Datacubes. In this folder two ocean data cubes were created: 1) The Particulate Organic Carbon (POC) Data Cube at 4 km and 2) the Oceanic Export Production, Phytoplankton Carbon and Primary Production Data Cube at 9 km.

Reviewer: ```suggestion
Information dedicated to the generation of the Ocean Datacubes. In this folder two ocean data cubes were created:   
1) The Particulate Organic Carbon (POC) Data Cube at 4 km and   
2) the Oceanic Export Production, Phytoplankton Carbon and Primary Production Data Cube at 9 km.
```

## Cubes generation process

The cubes generation process is divided in four phases:

### 1. Downloading required raw datasets

Datasets were downloaded via the source information provided in `cube4km.geojson` and `cube9km.geojson`. Datasets were downloaded in `.nc` format.

> To download the datasets, it is required to run the following files:

Reviewer: Please add a description, that deep-esdl-cube-gen python env needs to be used for executing the scripts. And how to do this.

```
# Download all files

> download-sources-4km.py

Reviewer: ```suggestion
cd inputs-collect/
python download-sources-4km.py
```

> download-sources-9km.py

Reviewer: ```suggestion
python download-sources-9km.py
```

```

> ### 2. Preprocessing datasets

Author: I required to convert each one of them into zarr first (multiple nc files per dataset to zarr). That's why they're saved first as multiple zarr and then concatenated and uploaded to s3 directly.

> ### 2. Preprocessing datasets

Reviewer: But this could also be done in S3 - you could write the zarrs into s3 and then concatenate them using the data saved to s3

> ### 2. Preprocessing datasets

Author: Ok, next time we do it like this :)

The raw data is preprocessed in order to create a single input `.zarr` cube per dataset. The preprocessing steps might involve time resampling and/or spatial resampling according to the dataset. The preprocessing code for each dataset is found at the `inputs-preprocess` folder. Note that some datasets have multiple parts according to the preprocessing steps that were applied:

```
# Oceanic Export Production

> inputs-preprocess/oceanic-export-production-to-zarr.py # Convert files to zarr

Reviewer: ```suggestion
cd ..
python inputs-preprocess/oceanic-export-production-to-zarr.py # Convert files to zarr
```

> inputs-preprocess/oceanic-export-production-to-s3.py # Upload full cube to the s3 bucket

Reviewer: ```suggestion
python inputs-preprocess/oceanic-export-production-to-s3.py # Upload full cube to the s3 bucket
```

# Particulate Organic Carbon

> inputs-preprocess/particulate-organic-carbon-to-zarr.py # Convert files to zarr

Reviewer: ```suggestion
python inputs-preprocess/particulate-organic-carbon-to-zarr.py # Convert files to zarr
```

# Phytoplankton Carbon

> inputs-preprocess/phytoplankton-carbon-to-zarr.py # Convert files to zarr

Reviewer: ```suggestion
python inputs-preprocess/phytoplankton-carbon-to-zarr.py # Convert files to zarr
```

> inputs-preprocess/phytoplankton-carbon-to-s3.py # Upload full cube to the s3 bucket

Reviewer: ```suggestion
python inputs-preprocess/phytoplankton-carbon-to-s3.py # Upload full cube to the s3 bucket
```

# Primary Production

> inputs-preprocess/primary-production-to-zarr.py # Convert files to zarr

Reviewer: ```suggestion
python inputs-preprocess/primary-production-to-zarr.py # Convert files to zarr
```

> inputs-preprocess/primary-production-to-s3.py # Upload full cube to the s3 bucket

Reviewer: ```suggestion
python inputs-preprocess/primary-production-to-s3.py # Upload full cube to the s3 bucket
```

```

### 3. Merging all datasets into a single cube

All `.zarr` cubes are loaded and merged by their coordinates and timesteps into a single `.zarr` cube.

```
# Merge all datasets

> output-merge/merge-4km.py

Reviewer: ```suggestion
python output-merge/merge-4km.py
```

> output-merge/merge-9km.py

Reviewer: ```suggestion
python output-merge/merge-9km.py
```

# Polar Datacube Generation

> Information dedicated to the generation of the Ocean Datacubes. In this folder two ocean data cubes were created:

Reviewer: ```suggestion
Information dedicated to the generation of the Ocean Datacubes. By using the following recipe two ocean data cubes can be reproduced:
```

1) The Particulate Organic Carbon (POC) Data Cube at 4 km and   
2) the Oceanic Export Production, Phytoplankton Carbon and Primary Production Data Cube at 9 km.

> The total storage required for the generation of this cube is 490 Gb.

Reviewer: ```suggestion
The total storage required for the generation of these cube is 490 Gb.
```

## Cubes generation process

The cubes generation process is divided in four phases:

### 1. Downloading required raw datasets

Datasets were downloaded into a folder named `data` in the user home path via the source information provided in `cube4km.geojson` and `cube9km.geojson`. Datasets were downloaded in `.nc` format.

To download the datasets, it is required to run the following files under the `cube-gen` python environment:
```
conda activate cube-gen
```

The files can be executed using the following commands:

```
# Download all files
cd inputs-collect/
python download-sources-4km.py
python download-sources-9km.py
```

### 2. Preprocessing datasets

> The raw data is preprocessed in order to create a single input `.zarr` cube per dataset. The preprocessing steps might involve time resampling and/or spatial resampling according to the dataset. The preprocessing code for each dataset is found at the `inputs-preprocess` folder. Note that some datasets have multiple parts according to the preprocessing steps that were applied:

Reviewer: ```suggestion
The raw data is preprocessed in order to create a single input `.zarr` cube per dataset. The preprocessing steps might involve time resampling and/or spatial resampling according to the dataset. The preprocessing code for each dataset is found at the `inputs-preprocess` folder. Note that some datasets have multiple parts according to the preprocessing steps that were applied:

**If you wish to execute the below scripts please ensure to adjust the `root` parameter in the scripts to point to a s3 bucket which you have read-write access to.**
```

cd ..
python inputs-preprocess/oceanic-export-production-to-zarr.py # Convert files to zarr
python inputs-preprocess/oceanic-export-production-to-s3.py # Upload full cube to the s3 bucket

# Particulate Organic Carbon
python inputs-preprocess/particulate-organic-carbon-to-zarr.py # Convert files to zarr

# Phytoplankton Carbon
python inputs-preprocess/phytoplankton-carbon-to-zarr.py # Convert files to zarr
python inputs-preprocess/phytoplankton-carbon-to-s3.py # Upload full cube to the s3 bucket

# Primary Production
python inputs-preprocess/primary-production-to-zarr.py # Convert files to zarr
python inputs-preprocess/primary-production-to-s3.py # Upload full cube to the s3 bucket
```

### 3. Merging all datasets into a single cube

All `.zarr` cubes are loaded and merged by their coordinates and timesteps into a single `.zarr` cube.

```
# Merge all datasets
python output-merge/merge-4km.py
python output-merge/merge-9km.py
```

### 4. Postprocessing

A patch of metadata is added to the final cubes using `xcube patch` and `output-postprocess/patch-4km.yaml`/`output-postprocess/patch-9km.yaml` by running:

```
# Merge all datasets
output-postprocess/patch.sh

> ```

Reviewer: ```suggestion
```

**If you wish to execute the above script please ensure to adjust the s3 bucket path to point to the bucket which you have read-write access to.**
```

> ```

Reviewer: It does not work well with the inline comment here. Could you please add: 

**If you wish to execute the above script please ensure to adjust the s3 bucket path to point to the bucket which you have read-write access to.**
